File: application/app/src/main/java/com/yahyeet/boardbook/presenter/matchfeed/matchdetail/MatchDetailPresenter.java
package com.yahyeet.boardbook.presenter.matchfeed.matchdetail;

import android.content.Context;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.yahyeet.boardbook.activity.IFutureInteractable;
import com.yahyeet.boardbook.activity.home.matchfeed.matchdetail.IMatchDetailActivity;
import com.yahyeet.boardbook.model.entity.MatchPlayer;
import com.yahyeet.boardbook.model.handler.MatchHandler;
import com.yahyeet.boardbook.presenter.BoardbookSingleton;
import com.yahyeet.boardbook.model.entity.Match;
import com.yahyeet.boardbook.presenter.FindOnePresenter;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Presenter for the match detail activity
 */
public class MatchDetailPresenter extends FindOnePresenter<Match, MatchHandler> {


	private IMatchDetailActivity matchDetailActivity;
	private MatchPlayerAdapter matchPlayerAdapter;

	private List<MatchPlayer> matchPlayers = new ArrayList<>();

	public MatchDetailPresenter(IMatchDetailActivity matchDetailActivity, String matchID) {
		super((IFutureInteractable) matchDetailActivity);
		this.matchDetailActivity = matchDetailActivity;

		findEntity(BoardbookSingleton.getInstance().getMatchHandler(), matchID,
			MatchHandler.generatePopulatorConfig(true, true));


	}


	private void setMatchDetailName() {
		matchDetailActivity.setGameName("Match of " + getEntity().getGame().getName());
	}


	/**
	 * Initiates the adapter for a matchplayerRecyclerView
	 *
	 * @param matchplayerRecyclerView view that will receive new adapter
	 * @param viewContext             structure of current view
	 */
	public void enableMatchplayerAdapter(RecyclerView matchplayerRecyclerView, Context viewContext) {

		RecyclerView.LayoutManager layoutManager = new GridLayoutManager(viewContext, 1);
		matchPlayerAdapter = new MatchPlayerAdapter(matchPlayers);

		matchplayerRecyclerView.setLayoutManager(layoutManager);
		matchplayerRecyclerView.setAdapter(matchPlayerAdapter);
	}

	@Override
	protected void onEntityFound(Match entity) {
		setMatchDetailName();

		matchPlayers.clear();
		matchPlayers.addAll(sortByWin(entity.getMatchPlayers()));

	}

	private List<MatchPlayer> sortByWin(List<MatchPlayer> unsortedPlayers) {

		return unsortedPlayers
			.stream()
			.sorted((left, right) -> -Boolean.compare(left.getWin(), right.getWin()))
			.collect(Collectors.toList());
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/presenter/matchfeed/matchdetail/MatchDetailPresenter.kt
package com.yahyeet.boardbook.presenter.matchfeed.matchdetail

import android.content.Context
import android.content.res.Resources
import android.os.Looper

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.yahyeet.boardbook.activity.IFutureInteractable
import com.yahyeet.boardbook.activity.home.matchfeed.matchdetail.IMatchDetailActivity
import com.yahyeet.boardbook.presenter.BoardbookSingleton
import com.yahyeet.boardbook.model.entity.Match


class MatchDetailPresenter(private val matchDetailActivity: IMatchDetailActivity, matchID: String) {
    private var matchPlayerAdapter: MatchPlayerAdapter? = null
    private var match: Match? = null

    init {


        if (matchDetailActivity is IFutureInteractable) {

            val futureDetail = matchDetailActivity as IFutureInteractable

            futureDetail.disableViewInteraction()
            BoardbookSingleton.getInstance().matchHandler.find(matchID).thenAccept { foundMatch ->
                match = foundMatch

                android.os.Handler(Looper.getMainLooper()).post {
                    setMatchDetailName()
                    matchDetailActivity.initiateMatchDetailList()
                    futureDetail.enableViewInteraction()
                }

            }.exceptionally { e ->
                e.printStackTrace()
                futureDetail.displayLoadingFailed()
                null
            }
        } else {
            throw IllegalArgumentException("Activity not instance of IFutureIntractable")
        }

    }


    private fun setMatchDetailName() {
        matchDetailActivity.setGameName("Game of " + match!!.game!!.name!!)
    }


    /**
     * Makes recyclerView to repopulate its matches with current data
     */
    fun updateMatchplayerAdapter() {
        matchPlayerAdapter!!.notifyDataSetChanged()
    }

    /**
     * Initiates the adapter for a matchplayerRecyclerView
     *
     * @param matchplayerRecyclerView view that will receive new adapter
     * @param viewContext             structure of current view
     */
    fun enableMatchplayerAdapter(matchplayerRecyclerView: RecyclerView, viewContext: Context, resources: Resources) {

        val layoutManager = GridLayoutManager(viewContext, 1)

        matchplayerRecyclerView.layoutManager = layoutManager

        matchPlayerAdapter = MatchPlayerAdapter(match!!.matchPlayers, resources)
        matchplayerRecyclerView.adapter = matchPlayerAdapter
    }


}
